packages/vite/src/node/config.ts

```diff
@@ -1081,25 +1104,73 @@ async function loadConfigFromBundledFile(
   }
 }
 
+const orders = ['pre', null, 'post'] as const
+
 async function runConfigHook(
   config: InlineConfig,
-  plugins: Plugin[],
-  configEnv: ConfigEnv
+  configEnv: ConfigEnv,
+  plugins: Plugin[]
 ): Promise<InlineConfig> {
-  let conf = config
-
-  for (const p of getSortedPluginsByHook('config', plugins)) {
-    const hook = p.config
-    const handler = hook && 'handler' in hook ? hook.handler : hook
-    if (handler) {
-      const res = await handler(conf, configEnv)
-      if (res) {
-        conf = mergeConfig(conf, res)
+  const calledPlugins = new Set<Plugin>()
+  const callConfigHandler = async (plugin: Plugin) => {
+    calledPlugins.add(plugin)
+    const handler =
+      typeof plugin.config !== 'function'
+        ? plugin.config!.handler!
+        : plugin.config!
+
+    const result = await handler(config, configEnv)
+    if (result) {
+      config = mergeConfig(config, result)
+    }
+  }
+
+  const enforce = (p: Plugin) =>
+    (p.config && ('handler' in p.config ? p.config.order : p.enforce)) || null
+
+  const isEligible = (p: Plugin, order: typeof orders[number]) => {
+    if (!p.config) {
+      return false
+    }
+    const handler = typeof p.config !== 'function' ? p.config.handler : p.config
+    if (!handler) {
+      return false
+    }
+    if (calledPlugins.has(p)) {
+      return false
+    }
+    return orders.indexOf(enforce(p)) <= orders.indexOf(order)
+  }
+
+  let needSort = false
+  for (const order of orders) {
+    for (let i = 0; i < plugins.length; i++) {
+      const plugin = plugins[i]
+      if (isEligible(plugin, order)) {
+        const oldPlugins = config.plugins!
+        await callConfigHandler(plugin)
+        const newPlugins = await flattenVitePlugins(
+          config.plugins!.slice(oldPlugins.length),
```

Comment:
Since `config.plugins` is immutable, the only way to add nested plugins is to return `{ plugin: [...] }` from the config hook. This makes it easy to find the nested plugins here, as `mergeConfig` guarantees they'll be at the end.